File: test-errors/src/factories/internal/metadata/emend_metadata_atomics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.emend_metadata_atomics = void 0;
const MetadataAtomic_1 = require("../../../schemas/metadata/MetadataAtomic");
const ArrayUtil_1 = require("../../../utils/ArrayUtil");
const emend_metadata_atomics = (meta) => {
    // ATOMICS
    for (const a of meta.atomics) {
        const index = meta.constants.findIndex((c) => c.type === a.type);
        if (index !== -1)
            meta.constants.splice(index, 1);
    }
    // BOOLEAN
    {
        const index = meta.constants.findIndex((c) => c.type === "boolean");
        if (index !== -1 && meta.constants[index].values.length === 2) {
            meta.constants.splice(index, 1);
            ArrayUtil_1.ArrayUtil.take(meta.atomics, (a) => a.type === "boolean", () => MetadataAtomic_1.MetadataAtomic.create({
                type: "boolean",
                tags: [],
            }));
        }
    }
    // TEMPLATE
    if (meta.templates.length &&
        meta.atomics.find((a) => a.type === "string") !== undefined)
        meta.templates.splice(0, meta.templates.length);
};
exports.emend_metadata_atomics = emend_metadata_atomics;

File: packages/core/src/factories/internal/metadata/emend_metadata_atomics.go
package metadata

func EmendMetadataAtomics(meta *MetadataSchema) {
	if meta == nil {
		return
	}
	filtered := meta.Constants[:0]
	for _, constant := range meta.Constants {
		remove := false
		for _, atomic := range meta.Atomics {
			if atomic.Type == constant.Type && !isNotPureAtomic(atomic) {
				remove = true
				break
			}
		}
		if !remove {
			filtered = append(filtered, constant)
		}
	}
	meta.Constants = filtered
	for index, constant := range meta.Constants {
		if constant.Type == "boolean" && len(constant.Values) == 2 {
			meta.Constants = append(meta.Constants[:index], meta.Constants[index+1:]...)
			tags := [][]MetadataTypeTag{}
			if len(constant.Values) != 0 {
				tags = constant.Values[0].Tags
			}
			takeAtomic(&meta.Atomics, MetadataAtomic{Type: "boolean", Tags: tags})
			break
		}
	}
	if len(meta.Templates) != 0 {
		for _, atomic := range meta.Atomics {
			if atomic.Type == "string" && !isNotPureAtomic(atomic) {
				meta.Templates = nil
				break
			}
		}
	}
}

func isNotPureAtomic(atomic MetadataAtomic) bool {
	if len(atomic.Tags) == 0 {
		return false
	}
	for _, row := range atomic.Tags {
		if len(row) == 0 {
			return false
		}
		for _, tag := range row {
			if tag.Validate == "" {
				return false
			}
		}
	}
	return true
}
